File: rest-v3/go/examples.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"time"
)

const apiBaseUrl = "https://api.foxbit.com.br"

func sign(method, path string, params map[string]string, body map[string]interface{}) (string, string) {
	var queryString string
	if params != nil {
		queryValues := url.Values{}
		for key, value := range params {
			queryValues.Add(key, value)
		}
		queryString = queryValues.Encode()
	}

	var rawBody []byte
	var err error
	if body != nil {
		rawBody, err = json.Marshal(body)
		if err != nil {
			log.Fatal("Error encoding body to JSON:", err)
		}
	}

	timestamp := strconv.FormatInt(time.Now().UnixMilli(), 10)
	preHash := timestamp + method + path + queryString + string(rawBody)
	fmt.Println("PreHash:", preHash)

	h := hmac.New(sha256.New, []byte(os.Getenv("FOXBIT_API_SECRET")))
	h.Write([]byte(preHash))
	signature := hex.EncodeToString(h.Sum(nil))
	fmt.Println("Signature:", signature)

	return signature, timestamp
}

func request(method, path string, params map[string]string, body map[string]interface{}) ([]byte, error) {
	fmt.Println("--------------------------------------------------")
	fmt.Println("Requesting:", method, path)
	signature, timestamp := sign(method, path, params, body)
	url := apiBaseUrl + path

	client := &http.Client{}
	var req *http.Request
	var err error

	if body != nil {
		jsonBody, _ := json.Marshal(body)
		req, err = http.NewRequest(method, url, bytes.NewBuffer(jsonBody))
	} else {
		req, err = http.NewRequest(method, url, nil)
	}

	if err != nil {
		log.Fatal("Error creating request:", err)
	}

	q := req.URL.Query()
	for key, value := range params {
		q.Add(key, value)
	}
	req.URL.RawQuery = q.Encode()

	req.Header.Add("X-FB-ACCESS-KEY", os.Getenv("FOXBIT_API_KEY"))
	req.Header.Add("X-FB-ACCESS-TIMESTAMP", timestamp)
	req.Header.Add("X-FB-ACCESS-SIGNATURE", signature)
	req.Header.Add("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		log.Fatal("Error on request:", err)
	}
	defer resp.Body.Close()

	bodyResp, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatal("Error reading response body:", err)
	}

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		fmt.Printf("HTTP Status Code: %d, Error Response Body: %s\n", resp.StatusCode, bodyResp)
		return nil, fmt.Errorf("request failed with status code %d", resp.StatusCode)
	}

	return bodyResp, nil
}

func createOrder(marketSymbol, side, orderType, price, quantity string) ([]byte, error) {
	order := map[string]interface{}{
		"market_symbol": marketSymbol,
		"side":          side,
		"type":          orderType,
		"price":         price,
		"quantity":      quantity,
	}
	response, err := request("POST", "/rest/v3/orders", nil, order)
	if err != nil {
		return nil, err
	}
	return response, nil
}

func cancelOrder(orderID string) ([]byte, error) {
	orderToCancel := map[string]interface{}{
		"type": "ID",
		"id":   orderID,
	}
	response, err := request("PUT", "/rest/v3/orders/cancel", nil, orderToCancel)
	if err != nil {
		return nil, err
	}
	return response, nil
}

func main() {
	// Get user information
	mePath := "/rest/v3/me"
	meResponse, err := request("GET", mePath, nil, nil)
	if err != nil {
		log.Fatal("Failed to get user information:", err)
	}
	fmt.Println("Response:", string(meResponse))

	// Create an order
	orderResponse, err := createOrder("btcbrl", "BUY", "LIMIT", "10.0", "0.0001")
	if err != nil {
		log.Fatal("Failed to create order:", err)
	}
	fmt.Println("Order Response:", string(orderResponse))

	// Get order information
	var orderData map[string]interface{}
	err = json.Unmarshal(orderResponse, &orderData)
	if err != nil {
		log.Fatal("Failed to parse order response:", err)
	}

	// Cancel an order
	orderID := orderData["id"].(string)
	cancelResponse, err := cancelOrder(orderID)
	if err != nil {
		log.Fatal("Failed to cancel order:", err)
	}
	fmt.Println("Cancel Response:", string(cancelResponse))
}

File: rest-v3/python/examples.py
import os
import json
import requests
import time
import hmac
import hashlib
from urllib.parse import urlencode

api_key = os.getenv('FOXBIT_API_KEY')
api_secret = os.getenv('FOXBIT_API_SECRET')
api_base_url = 'https://api.foxbit.com.br'

def sign(method, path, params, body):
    queryString = ''
    if params:
        queryString = urlencode(params)

    rawBody = ''
    if body:
        rawBody = json.dumps(body)

    timestamp = str(int(time.time() * 1000))
    preHash = f"{timestamp}{method.upper()}{path}{queryString}{rawBody}"
    print('PreHash:', preHash)
    signature = hmac.new(api_secret.encode(), preHash.encode(), hashlib.sha256).hexdigest()
    print('Signature:', signature)

    return signature, timestamp

def request(method, path, params, body):
    print('--------------------------------------------------')
    print('Requesting:', method, path)
    signature, timestamp = sign(method, path, params, body)
    url = f"{api_base_url}{path}"
    headers = {
        'X-FB-ACCESS-KEY': api_key,
        'X-FB-ACCESS-TIMESTAMP': timestamp,
        'X-FB-ACCESS-SIGNATURE': signature,
        'Content-Type': 'application/json',
    }

    try:
        response = requests.request(method, url, params=params, json=body, headers=headers)
        response.raise_for_status()
        return response.json()
    except requests.HTTPError as http_err:
        print(f"HTTP Status Code: {http_err.response.status_code}, Error Response Body:", http_err.response.json())
        raise
    except Exception as err:
        print(f"An error occurred: {err}")
        raise

if __name__ == '__main__':
    try:
        print('FOXBIT_API_KEY:', api_key)

        # Get user info
        meResponse = request('GET', '/rest/v3/me', None, None)
        print('Response:', meResponse)

        # Create a new order
        order = {
            'market_symbol': 'btcbrl',
            'side': 'BUY',
            'type': 'LIMIT',
            'price': '10.0',
            'quantity': '0.0001',
        }
        orderResponse = request('POST', '/rest/v3/orders', None, order)
        print('Response:', orderResponse)

        time.sleep(2)

        # Get active orders
        ordersParams = {
            'market_symbol': 'btcbrl',
            'state': 'ACTIVE'
        }
        ordersResponse = request('GET', '/rest/v3/orders', ordersParams, None)
        print('Response:', ordersResponse)

        # Cancel the order
        orderToCancel = {
            'type': 'ID',
            'id': orderResponse['id']
        }
        cancelResponse = request('PUT', '/rest/v3/orders/cancel', None, orderToCancel)
        print('Response:', cancelResponse)

    except Exception as e:
        print('Failed to process request.', str(e))
